### PR TITLE
Bug fix/refactor query optimizer part2

### PR DESCRIPTION
**Pull request by jsteemann**

### CODE REVIEW

**Comment by jsteemann** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/2044/

**Comment by jsteemann** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/2045/

**Comment by mpoeter** [comment]
is it really necessary to copy the dependencies vector in all these cases? There are some places with a comment "intentional copy", but for the other cases I figured it should be fine to change `getDependencies` to return a const ref instead of a copy.

**Comment by jsteemann** [comment]
The problem is that in the cases I changed there were crashes because of assertions failures.
`getDependencies()` was now returning a reference to the dependencies, which is fine. However, directly afterwards in the places I changed it there were other calls that modified the execution plan, e.g. by inserting and removing dependencies. This also modified the value we got via the call to `getDependencies()` before. There was no multi-threading issue, but simply the generated execution plans became invalid.
The problem only occurred in the cluster, single-server mode was fine.

**Comment by jsteemann** [comment]
I can put some "intentional copy" above the places I changed in order to indicate that the copy here is also intentional.

File: arangod/Aql/OptimizerRules.cpp [hunk 1]
```diff
@@ -4214,7 +4218,7 @@ void arangodb::aql::collectInClusterRule(Optimizer* opt,
 
     // found a node we need to replace in the plan
 
-    auto const& deps = node->getDependencies();
+    auto const deps = node->getDependencies();
     TRI_ASSERT(deps.size() == 1);
```
Comment (mpoeter) [inline arangod/Aql/OptimizerRules.cpp]: Don't need to copy here as deps is only used in the assertion.

File: arangod/Aql/OptimizerRules.cpp [hunk 2]
```diff
@@ -4703,7 +4709,9 @@ void arangodb::aql::distributeSortToClusterRule(
   bool modified = false;
 
   for (auto& n : nodes) {
-    auto const& remoteNodeList = n->getDependencies();
+    // intentional copy of the dependencies, as we will be modifying 
+    // dependencies later on
+    auto const remoteNodeList = n->getDependencies();
     auto gatherNode = ExecutionNode::castTo<GatherNode*>(n);
     TRI_ASSERT(remoteNodeList.size() > 0);
     auto rn = remoteNodeList[0];
```
Comment (mpoeter) [inline arangod/Aql/OptimizerRules.cpp]: Is this copy really necessary here? As far as I can see remoteNodeList is only used these 4 lines.

**Comment by jsteemann** [comment]
I think it is safe to use references in the two above places you mentioned.
I made the changes and started yet another test.
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/2049/

**Pull request merged**